py/modsys.c

```diff
@@ -208,13 +208,42 @@ static MP_DEFINE_CONST_FUN_OBJ_1(mp_sys_getsizeof_obj, mp_sys_getsizeof);
 #endif
 
 #if MICROPY_PY_SYS_ATEXIT
+typedef struct _m_atexit_node_t {
+    struct _m_atexit_node_t *prev;
+    struct _m_atexit_node_t *next;
+    mp_obj_t fn;
+} m_atexit_node_t;
+
 // atexit(callback): Callback is called when sys.exit is called.
 static mp_obj_t mp_sys_atexit(mp_obj_t obj) {
-    mp_obj_t old = MP_STATE_VM(sys_exitfunc);
-    MP_STATE_VM(sys_exitfunc) = obj;
-    return old;
+    m_atexit_node_t *node = m_malloc(sizeof(m_atexit_node_t));
+    if (MP_STATE_VM(sys_exitfunc) != mp_const_none) {
+        MP_STATE_VM(sys_exitfunc)->prev = node;
+    }
+    node->fn = obj;
+    node->prev = NULL;
+    node->next = MP_STATE_VM(sys_exitfunc);
+    MP_STATE_VM(sys_exitfunc) = node;
+    return obj;
 }
 static MP_DEFINE_CONST_FUN_OBJ_1(mp_sys_atexit_obj, mp_sys_atexit);
+
+void mp_sys_atexit_execute(void) {
+    // walk through the linked list and execute each function
+    // Beware, the sys.settrace callback should be disabled before running sys.atexit.
+    while (MP_STATE_VM(sys_exitfunc) != mp_const_none) {
+        if (mp_obj_is_callable(MP_STATE_VM(sys_exitfunc)->fn)) {
+            nlr_buf_t nlr;
+            if (nlr_push(&nlr) == 0) {
+                mp_call_function_0(MP_STATE_VM(sys_exitfunc)->fn);
+            } else {
+                // Uncaught exception: print it out.
```

Comment:
In CPython a SystemExit would still cause an actual exit, is it worth adding that?

Reply:
These are currently only being run during an exit / soft-restart so I wouldn't think the handling should change? 
Though maybe the exception _shouldn't_ be printed out if it's a `SystemExit`

Reply:
> so I wouldn't think the handling should change?

It is actually changing: for the unix port an exception called during the atexit callback would result in exit code 1 (because of an 'FATAL: uncaught NLR', maybe not ideal, but at least it's a non-zero exit code and that is what matters). But with this implementation the exception would be swallowed (and printed). That should not change imo and this should remain CPython-compatible in the sense that an exception during atexit is treated as an exception. Note CPython does [this](https://docs.python.org/3/library/atexit.html):

> If an exception is raised during execution of the exit handlers, a traceback is printed (unless SystemExit is raised) and the exception information is saved. After all exit handlers have had a chance to run, the last exception to be raised is re-raised.

But given your question, I assume raising an exception here would actually change behavior?


>  Though maybe the exception shouldn't be printed out if it's a SystemExit

That's indeed what CPython does.

Reply:
Ah, that detail helps thanks! The filter on whether to print the exception is easy enough to implement certainly. Maybe then just overriding the exit code in the exception handler here is enough to broadly match cpython, as opposed to keeping a copy of the exception details to pass them onto any default handler.

Reply:
> Maybe then just overriding the exit code in the exception handler here is enough to broadly match cpython

Probably, but in the end it would be nicer to fix this properly:


> as opposed to keeping a copy of the exception details to pass them onto any default handler.

Ideally the only thing which would be needed is to store `nlr.ret_val` then if not NULL raise it again after having called all handlers. Should not be an issue code-size-wise. Passing it onto the default handler would then happen automatically because the main code knows how to do that; however, this requires that `mp_sys_atexit_execute` is being called within `execute_from_lexer` (or whatever the counterpart is named for the various ports).

This is not the case for the unix port now, nor for the other ports. I don't know why though, and all in all this is maybe not very logical/expected; e.g. for the stm32 port it runs after deinitializing pretty much all of the hardware: is there a specific reason you did it that way? Would it not be more sound if there's not many restrictions to what can run in the atexit handler i.e. can still use hardware if it chooses to etc?